File: internal/wgconfig/wgconfig.go
package wgconfig

import (
	"bufio"
	"bytes"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

type ApplyRequest struct {
	ConfigType    string
	Interface     string
	PeerPublicKey string
	EndpointHost  string
	EndpointPort  int
	ConfigPath    string
	ReloadMethod  string
	DryRun        bool
}

type ApplyResult struct {
	Changed bool   `json:"changed"`
	Message string `json:"message"`
}

func ApplyEndpoint(req ApplyRequest) (ApplyResult, error) {
	endpoint := fmt.Sprintf("%s:%d", req.EndpointHost, req.EndpointPort)
	var result ApplyResult
	var err error
	switch req.ConfigType {
	case "wg_conf":
		path := req.ConfigPath
		if path == "" {
			path = filepath.Join("/etc/wireguard", req.Interface+".conf")
		}
		result, err = UpdateWGConf(path, req.PeerPublicKey, endpoint, req.DryRun)
	case "openwrt_uci":
		result, err = UpdateOpenWrtUCI(req.Interface, req.PeerPublicKey, req.EndpointHost, req.EndpointPort, req.DryRun)
	case "runtime":
		result, err = WGSetEndpoint(req.Interface, req.PeerPublicKey, endpoint, req.DryRun)
	default:
		err = fmt.Errorf("unsupported config_type: %s", req.ConfigType)
	}
	if err != nil {
		return ApplyResult{}, err
	}
	if !req.DryRun && req.ReloadMethod != "" && req.ReloadMethod != "none" {
		if err := ReloadInterface(req.Interface, req.ReloadMethod); err != nil {
			return ApplyResult{}, err
		}
	}
	return result, nil
}

func UpdateWGConf(path, peerPublicKey, endpoint string, dryRun bool) (ApplyResult, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return ApplyResult{}, err
	}
	lines := splitKeepNewline(raw)
	inPeer := false
	matchingPeer := false
	endpointIndex := -1
	insertIndex := -1

	for i, line := range lines {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "[") && strings.HasSuffix(trimmed, "]") {
			if inPeer && matchingPeer {
				insertIndex = i
				break
			}
			inPeer = strings.EqualFold(trimmed, "[Peer]")
			matchingPeer = false
			endpointIndex = -1
			continue
		}
		if !inPeer || strings.HasPrefix(trimmed, "#") || !strings.Contains(trimmed, "=") {
			continue
		}
		parts := strings.SplitN(trimmed, "=", 2)
		key := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(parts[1])
		switch {
		case strings.EqualFold(key, "PublicKey") && value == peerPublicKey:
			matchingPeer = true
		case strings.EqualFold(key, "Endpoint") && matchingPeer:
			endpointIndex = i
		}
	}
	if inPeer && matchingPeer && insertIndex == -1 {
		insertIndex = len(lines)
	}
	if !matchingPeer {
		return ApplyResult{}, errors.New("peer public key not found in wg config")
	}

	newLine := "Endpoint = " + endpoint + "\n"
	if endpointIndex >= 0 {
		if strings.TrimSpace(lines[endpointIndex]) == strings.TrimSpace(newLine) {
			return ApplyResult{Changed: false, Message: "endpoint already up to date"}, nil
		}
		lines[endpointIndex] = newLine
	} else {
		lines = append(lines[:insertIndex], append([]string{newLine}, lines[insertIndex:]...)...)
	}
	if dryRun {
		return ApplyResult{Changed: true, Message: "dry-run: wg.conf would be updated"}, nil
	}
	backup := fmt.Sprintf("%s.bak.%d", path, time.Now().Unix())
	if err := os.WriteFile(backup, raw, 0o600); err != nil {
		return ApplyResult{}, err
	}
	return ApplyResult{Changed: true, Message: "updated " + path}, os.WriteFile(path, []byte(strings.Join(lines, "")), 0o600)
}

func UpdateOpenWrtUCI(interfaceName, peerPublicKey, endpointHost string, endpointPort int, dryRun bool) (ApplyResult, error) {
	if dryRun {
		return ApplyResult{Changed: true, Message: "dry-run: would update OpenWrt " + interfaceName + " peer endpoint"}, nil
	}
	section, err := findOpenWrtPeerSection(interfaceName, peerPublicKey)
	if err != nil {
		return ApplyResult{}, err
	}
	if err := run("uci", "set", fmt.Sprintf("network.%s.endpoint_host=%s", section, endpointHost)); err != nil {
		return ApplyResult{}, err
	}
	if err := run("uci", "set", fmt.Sprintf("network.%s.endpoint_port=%d", section, endpointPort)); err != nil {
		return ApplyResult{}, err
	}
	if err := run("uci", "commit", "network"); err != nil {
		return ApplyResult{}, err
	}
	return ApplyResult{Changed: true, Message: "updated OpenWrt section " + section}, nil
}

func WGSetEndpoint(interfaceName, peerPublicKey, endpoint string, dryRun bool) (ApplyResult, error) {
	if dryRun {
		return ApplyResult{Changed: true, Message: "dry-run: runtime endpoint would be updated"}, nil
	}
	if err := run("wg", "set", interfaceName, "peer", peerPublicKey, "endpoint", endpoint); err != nil {
		return ApplyResult{}, err
	}
	return ApplyResult{Changed: true, Message: "updated runtime endpoint"}, nil
}

func ReloadInterface(interfaceName, method string) error {
	switch method {
	case "ifup":
		_ = run("ifdown", interfaceName)
		return run("ifup", interfaceName)
	case "wg-quick-restart":
		return run("systemctl", "restart", "wg-quick@"+interfaceName)
	case "network-reload":
		return run("/etc/init.d/network", "reload")
	default:
		return fmt.Errorf("unsupported reload_method: %s", method)
	}
}

func StopInterface(interfaceName, method string) error {
	switch method {
	case "ifup", "openwrt":
		return run("ifdown", interfaceName)
	case "wg-quick":
		return run("wg-quick", "down", interfaceName)
	case "systemd":
		return run("systemctl", "stop", "wg-quick@"+interfaceName)
	default:
		return fmt.Errorf("unsupported wireguard_control_method: %s", method)
	}
}

func StartInterface(interfaceName, method string) error {
	switch method {
	case "ifup", "openwrt":
		return run("ifup", interfaceName)
	case "wg-quick":
		return run("wg-quick", "up", interfaceName)
	case "systemd":
		return run("systemctl", "start", "wg-quick@"+interfaceName)
	default:
		return fmt.Errorf("unsupported wireguard_control_method: %s", method)
	}
}

func findOpenWrtPeerSection(interfaceName, peerPublicKey string) (string, error) {
	cmd := exec.Command("uci", "show", "network")
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	prefix := "network.@wireguard_" + interfaceName + "["
	scanner := bufio.NewScanner(bytes.NewReader(out))
	for scanner.Scan() {
		line := scanner.Text()
		if !strings.HasPrefix(line, prefix) || !strings.Contains(line, ".public_key=") {
			continue
		}
		left, right, _ := strings.Cut(line, "=")
		value := strings.Trim(right, "'\"")
		if value == peerPublicKey {
			return strings.TrimPrefix(strings.TrimSuffix(left, ".public_key"), "network."), nil
		}
	}
	if err := scanner.Err(); err != nil {
		return "", err
	}
	return "", errors.New("peer public key not found in OpenWrt UCI network config")
}

func run(name string, args ...string) error {
	cmd := exec.Command(name, args...)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("%s %s failed: %w: %s", name, strings.Join(args, " "), err, strings.TrimSpace(string(out)))
	}
	return nil
}

func splitKeepNewline(raw []byte) []string {
	if len(raw) == 0 {
		return nil
	}
	parts := strings.SplitAfter(string(raw), "\n")
	if parts[len(parts)-1] == "" {
		return parts[:len(parts)-1]
	}
	return parts
}

File: wgnh/wgconfig.py
from __future__ import annotations

import shutil
import subprocess
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class ApplyEndpointResult:
    changed: bool
    message: str


def apply_endpoint(
    *,
    config_type: str,
    interface: str,
    peer_public_key: str,
    endpoint_host: str,
    endpoint_port: int,
    config_path: str = "",
    reload_method: str = "none",
    dry_run: bool = False,
) -> ApplyEndpointResult:
    if config_type == "wg_conf":
        result = update_wg_conf(
            Path(config_path or f"/etc/wireguard/{interface}.conf"),
            peer_public_key,
            f"{endpoint_host}:{endpoint_port}",
            dry_run=dry_run,
        )
    elif config_type == "openwrt_uci":
        result = update_openwrt_uci(
            interface,
            peer_public_key,
            endpoint_host,
            endpoint_port,
            dry_run=dry_run,
        )
    elif config_type == "runtime":
        result = wg_set_endpoint(interface, peer_public_key, f"{endpoint_host}:{endpoint_port}", dry_run=dry_run)
    else:
        raise ValueError(f"unsupported config_type: {config_type}")

    if not dry_run and reload_method != "none":
        reload_interface(interface, reload_method)
    return result


def update_wg_conf(path: Path, peer_public_key: str, endpoint: str, dry_run: bool = False) -> ApplyEndpointResult:
    lines = path.read_text(encoding="utf-8").splitlines(keepends=True)
    in_peer = False
    matching_peer = False
    endpoint_line_index: int | None = None
    insert_index: int | None = None

    for idx, line in enumerate(lines):
        stripped = line.strip()
        if stripped.startswith("[") and stripped.endswith("]"):
            if in_peer and matching_peer:
                insert_index = idx
                break
            in_peer = stripped.lower() == "[peer]"
            matching_peer = False
            endpoint_line_index = None
            continue

        if not in_peer or stripped.startswith("#") or "=" not in stripped:
            continue

        key, value = [part.strip() for part in stripped.split("=", 1)]
        if key.lower() == "publickey" and value == peer_public_key:
            matching_peer = True
        elif key.lower() == "endpoint" and matching_peer:
            endpoint_line_index = idx

    if in_peer and matching_peer and insert_index is None:
        insert_index = len(lines)

    if not matching_peer:
        raise ValueError("peer public key not found in wg config")

    new_line = f"Endpoint = {endpoint}\n"
    if endpoint_line_index is not None:
        if lines[endpoint_line_index].strip() == new_line.strip():
            return ApplyEndpointResult(False, "endpoint already up to date")
        lines[endpoint_line_index] = new_line
    else:
        lines.insert(insert_index if insert_index is not None else len(lines), new_line)

    if dry_run:
        return ApplyEndpointResult(True, "dry-run: wg.conf would be updated")

    backup = path.with_suffix(path.suffix + ".bak")
    shutil.copy2(path, backup)
    path.write_text("".join(lines), encoding="utf-8")
    return ApplyEndpointResult(True, f"updated {path}")


def update_openwrt_uci(
    interface: str,
    peer_public_key: str,
    endpoint_host: str,
    endpoint_port: int,
    dry_run: bool = False,
) -> ApplyEndpointResult:
    if dry_run:
        return ApplyEndpointResult(True, f"dry-run: would update OpenWrt {interface} peer endpoint")
    section = find_openwrt_peer_section(interface, peer_public_key)
    run_checked(["uci", "set", f"network.{section}.endpoint_host={endpoint_host}"])
    run_checked(["uci", "set", f"network.{section}.endpoint_port={endpoint_port}"])
    run_checked(["uci", "commit", "network"])
    return ApplyEndpointResult(True, f"updated OpenWrt section {section}")


def find_openwrt_peer_section(interface: str, peer_public_key: str) -> str:
    raw = run_checked(["uci", "show", "network"], capture=True)
    prefix = f"network.@wireguard_{interface}["
    current = ""
    for line in raw.splitlines():
        if line.startswith(prefix) and ".public_key=" in line:
            section = line.split(".public_key=", 1)[0].removeprefix("network.")
            value = line.split("=", 1)[1].strip("'\"")
            if value == peer_public_key:
                current = section
                break
    if not current:
        raise ValueError("peer public key not found in OpenWrt UCI network config")
    return current


def wg_set_endpoint(interface: str, peer_public_key: str, endpoint: str, dry_run: bool = False) -> ApplyEndpointResult:
    if dry_run:
        return ApplyEndpointResult(True, "dry-run: runtime endpoint would be updated")
    run_checked(["wg", "set", interface, "peer", peer_public_key, "endpoint", endpoint])
    return ApplyEndpointResult(True, "updated runtime endpoint")


def reload_interface(interface: str, reload_method: str) -> None:
    if reload_method == "ifup":
        subprocess.run(["ifdown", interface], check=False)
        run_checked(["ifup", interface])
    elif reload_method == "wg-quick-restart":
        run_checked(["systemctl", "restart", f"wg-quick@{interface}"])
    elif reload_method == "network-reload":
        run_checked(["/etc/init.d/network", "reload"])
    else:
        raise ValueError(f"unsupported reload_method: {reload_method}")


def run_checked(args: list[str], capture: bool = False) -> str:
    proc = subprocess.run(
        args,
        check=True,
        text=True,
        stdout=subprocess.PIPE if capture else None,
        stderr=subprocess.PIPE,
    )
    return proc.stdout if capture else ""
